docs/proposals/20231227-koordlet-resctrl-qos-enhance.md

```diff
@@ -0,0 +1,468 @@
+
+}
+
+```
+
+**podsInformer**
+
+podsInformer will add a new interface to get all pods' taskids, Resctl runtime hook Reconciler will register callback to consume these information and write new taskIds to resctrl contrl group and monitor group.
+
+```go
+
+type PodMeta struct {
+    Pod       *corev1.Pod
+    CgroupDir string
+    // Add new field, todo: memory usage
+    ContainerTaskIds map[string][]uint64
+}
+
+func readPodTaskIds(cgroup string) []uint64 {
+
+}
+
+func (s *podsInformer) syncPods() error {
+    ...
+    for _, pod := range podList.Items {
+        podMeta := &statesinformer.PodMeta{
+            Pod:       pod.DeepCopy(),
+            CgroupDir: genPodCgroupParentDir(&pod), 
+            ContainerTaskIds: make([]uint64)
+        }
+        newPodMap[string(pod.UID)] = podMeta
+        // Add ContainerTaskIds to podMeta
+        // record pod container metrics
+        recordPodResourceMetrics(podMeta)
+    }
+    ...
+}
+```
+
+**Reconciler**
+
+Reconciler will help guarantee eventual consistency of Resctrl configuration. It will reconcile all QoS class resctrl config based on NodeSLO. For pod level, it will reconcile all pods resctrl config based on their annotations. 
+
+
+```go
+func (c *reconciler) reconcilePodCgroup(stopCh <-chan struct{}) {
+    // TODO refactor pod reconciler, inotify watch corresponding cgroup file and update only when receive modified event
+    // new watcher will be added with new pod created, and deleted with pod destroyed
+    for {
+        select {
+            // add new pod event handler
+            case <-c.PodAdded:
+                // add resctrl group for specific pod
+            case <-c.podUpdated:
+                podsMeta := c.getPodsMeta()
+                // call Resctrl runtime hook reconcilerFn to write new taskids to cgroup
+            // add new pod event handler
+            case <-c.PodRemoved:
+                // remove resctrl group for specific pod
+        ......
+        }
+        }
+    }
+}
+```
+
+#### Resctrl Metrics Collector
+
+Currently, we only retrieve QoS class level resctrl data like LLC and MBA. For further pod level monitor data, we will iterate all pods and get pod level metric.
+1. collector check whether resctrl file system supported and mounted
+2. iterate all QoS class monitor group in resctrl file system and read data from it and save data to DB
+4. iterate all Pods monitor group in resctrl file system and read data from it and save data to DB
+```go
+func (p *ResourceCtrlCollector) collectQoSClassResctrlResUsed() {
+    // NodeSLO, get all class level Resctrl usage
+    for class := range classes {
+        resctrl, err := GetResctrlUsage(classCgroupPath)
+    }
+}
+
+func (p *ResourceCtrlCollector) collectQoSClassResctrlResUsed() {
+    // Pods, get all specific pod Resctrl usage
+    for pod := range pods {
+        resctrl, err := GetResctrlUsage(podCgroupPath)
+    }
+}
+```
+
```

Comment:
we need a chapter about monitoring and observability, details can be left to future.